ext/opentelemetry-ext-opentracing-shim/src/opentelemetry/ext/opentracing_shim/__init__.py

```diff
@@ -0,0 +1,288 @@
+
+    When calling the initializer while passing a `SpanShim` object in the
+    `span` argument, the `ScopeShim` is initialized as a regular `Scope`
+    object.
+
+    When calling the initializer while passing a context manager *generator* in
+    the `span_cm` argument (as returned by the `use_span()` method of
+    OpenTelemetry `Tracer` objects,  the resulting `ScopeShim` becomes
+    usable as a context manager (using `with` statements).
+
+    It is necessary to have both ways for constructing `ScopeShim`
+    objects because in some cases we need to create the object from a context
+    manager, in which case our only way of retrieving a `Span` object is by
+    calling the `__enter__()` method on the context manager, which makes the
+    span active in the OpenTelemetry tracer; whereas in other cases we need to
+    accept a `SpanShim` object and wrap it in a `ScopeShim`.
+    """
+
+    def __init__(self, manager, span=None, span_cm=None):
+        super().__init__(manager, span)
+        self._span_cm = span_cm
+
+        # If a span context manager is provided, extract the `Span` object from
+        # it, wrap the extracted span and save it as an attribute.
+        if self._span_cm is not None:
+            otel_span = self._span_cm.__enter__()
+            self._span = SpanShim(
+                self._manager.tracer, otel_span.get_context(), otel_span
+            )
+
+    def close(self):
+        if self._span_cm is not None:
+            # We don't have error information to pass to `__exit__()` so we
+            # pass `None` in all arguments. If the OpenTelemetry tracer
+            # implementation requires this information, the `__exit__()` method
+            # on `opentracing.Scope` should be overridden and modified to pass
+            # the relevant values to this `close()` method.
+            self._span_cm.__exit__(None, None, None)
+        else:
+            self._span.unwrap().end()
+        # TODO: Set active span on OpenTelemetry tracer.
+        # https://github.com/open-telemetry/opentelemetry-python/issues/161#issuecomment-534136274
+
+
+class ScopeManagerShim(opentracing.ScopeManager):
+    def __init__(self, tracer):
+        # The only thing the `__init__()` method on the base class does is
+        # initialize `self._noop_span` and `self._noop_scope` with no-op
+        # objects. Therefore, it doesn't seem useful to call it.
+        # pylint: disable=super-init-not-called
+        self._tracer = tracer
+
+    def activate(self, span, finish_on_close):
+        span_cm = self._tracer.unwrap().use_span(
+            span.unwrap(), end_on_exit=finish_on_close
+        )
+        return ScopeShim(self, span_cm=span_cm)
+
+    @property
+    def active(self):
+        span = self._tracer.unwrap().get_current_span()
+        if span is None:
+            return None
+
+        wrapped_span = SpanShim(self._tracer, span.get_context(), span)
+        return ScopeShim(self, span=wrapped_span)
```

Comment:
This is an imperfect solution: The returned scope will always end the span. Bt that's also a really tricky problem.

Probably we should store the ScopeShim by monkey-patching it onto the otel_span. But even if we do that, the problem of what to do if an OpenTelemetry Span is active with no matching OpenTracing scope remains. I think we could only solve that by keeping track of the active context manager returned from `use_span` in the OpenTelemetry tracer (note that as implemented currently, a use_span context manager is only activtated by calling `__enter__`, not just by calling `use_span`.

Reply:
Right. How do you suggest we proceed with this @Oberon00? Add a TODO and maybe a follow up issue? Solve this before merging?
We already have a TODO for a [similar](https://github.com/open-telemetry/opentelemetry-python/pull/211/files#diff-cae01bbcb3091a3af5259b1611fa0157R176) case. Do you see the two as separate cases?

Reply:
I'd say we can just add a TODO here for now. I would assume that most of the time when you call `active` you are just interested in the Span behind the Scope and you wouldn't use the Scope in a `with` block or call `__exit__` or `close` on it, so it shouldn't matter in the usual case.

Reply:
Added a TODO.